File: C#/325. Maximum Size Subarray Sum Equals k.cs
/*
 * Link: https://leetcode.com/problems/maximum-size-subarray-sum-equals-k/submissions/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public int MaxSubArrayLen(int[] nums, int k) {
        var map = new Dictionary<int, int>();
        var sum = 0;
        var maxLen = 0;
        
        map.Add(0, 0);
        for(var i = 0; i < nums.Length; i++) {
            sum += nums[i];
            
            if (map.ContainsKey(sum - k)) {
                maxLen = Math.Max(maxLen, i + 1 - map[sum - k]);   
            }
            
            if (!map.ContainsKey(sum))map.Add(sum, i + 1);
        }
        
        return maxLen;
    }
}

File: Python/325. Maximum Size Subarray Sum Equals k.py
class Solution(object):
    def maxSubArrayLen(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        res, sum = 0, 0
        map = defaultdict(list)
        map[0] = -1
        for i in range(len(nums)):
            n = nums[i]
            sum += n
            if sum - k in map:
                res = max(res, i - map[sum - k])
            if sum not in map:
                map[sum] = i
        return res